### DISCUSSION TITLE
Repeat :action n times

### DISCUSSION THREAD

**Discussion by chylex** [JetBrains/ideavim#381] [Q&A]
Hi, is there a way to create a mapping that can be repeated if a number precedes it, for example

```
map gj :action EditorCloneCaretBelow<CR>
```

where `5gj` runs "clone caret below" 5 times?

**Comment by lippfi** [comment]
In Vim it would be possible if you used `v:count` and a custom vimscript function (correct me if there is an easier way to do it). But in IdeaVim the `v:` scope is not supported yet
For now the only solution I see is kinda tricky:
1. First of all you need somehow get the number passed to your mapping (`v:count`'s analog) and store it in a variable (let's call it `ideavim_prefix_number`)
btw range functions will be available in IdeaVim 1.8.0, but now you can test it in IdeaVim 1.8.0-eap.1 (https://github.com/JetBrains/ideavim#get-early-access)
```
let ideavim_prefix_number = 0
function! GetNumberPassedToMapping() range
  let g:ideavim_prefix_number = a:lastline - a:firstline + 1
endfunction
```
2. Then add a function that would do the action required number of times:
```
function! DoMyMappingNTimes(n)
  let i = 0
  while i < a:n
    execute ':action EditorCloneCaretBelow'
    let i += 1
  endwhile
endfunction
```
3. Combine this two functions to a single mapping:
```
nnoremap nn :call GetNumberPassedToMapping()<CR>:<C-U>call DoMyMappingNTimes(g:ideavim_prefix_number)<CR>
```

**Reply by chylex** [reply]
Thanks, that looks like a lot :D

I have about a hundred actions in my .ideavimrc, so I'm wondering if there could be a simpler solution. I noticed IdeaVim internally gets access to the range when I type a number and follow by an action keybind.

Would there be interest if I PR say, `:repeatableaction / <RepeatableAction>` that can process the number prefix automatically?

**Comment by lippfi** [comment]
I don't think that it's a good idea because customization is the thing that makes vim so special. Vim users can create `RepeatebleAction` command by their own, there is no need to add a new command in a PR
```
function! ExecuteActionNTimes(action) range
  let number_before_mapping = a:lastline - a:firstline + 1
  let i = 0
  while i < number_before_mapping
    execute 'action ' .. a:action
    let i += 1
  endwhile
endfunction

command! -nargs=1 -range RepeatebleAction :<line1>,<line2>call ExecuteActionNTimes(<q-args>)
```
So the mapping would look like this `nnoremap nn :RepeatebleAction EditorCloneCaretBelow<CR>`

`-range` attribute is not supported in `command` command, but if you did a PR with the attribute, it would be really cool

**Comment by lippfi** [comment]
IdeaVim also provides another way to map actions - `map gj <Action>(EditorCloneCaretBelow)`, which should support number before mapping, but it is broken at the moment (https://youtrack.jetbrains.com/issue/VIM-2431)
We will do our best to fix `map .. <Action>` asap

**Reply by chylex** [reply]
Hi, are there any updates on fixing the number before `<Action>` mapping?

**Reply by lippfi** [reply]
Hi, we are doing some huge refactoring to support the new Fleet IDE, so no updates, sorry

**Reply by DasOhmoff** [reply]
Hey, any news on this yet? This seems to be a very essential feature that would help out the community a lot

**Reply by lippfi** [reply]
Still no updates. It's not something that is requested by a lot of users and we try to address more requested issues first. You can upvote the ticket (https://youtrack.jetbrains.com/issue/VIM-2431) in YouTrack and increase its priority
PRs for addressing the ticket above or adding the '-range' option to the command are very welcome

**Reply by chylex** [reply]
I worked around the issue in my fork. I don't know if I'd recommend it because I occasionally have issues in my fork and haven't confirmed which changes caused them, but it works well enough for me. https://github.com/chylex/IntelliJ-IdeaVim/commit/ee178e58d0c83b3f844d630b7a114cc8e9ff073e

**Reply by DasOhmoff** [reply]
@lippfi I see. Well, thats too sad, but thanks for the info.

@chylex Ah, thats very good to know. Maybe thats all that's needed to make this work
